miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/api/ApiClient.kt

```diff
@@ -1,4 +1,81 @@
 package com.rakuten.tech.mobile.miniapp.api
 
-@SuppressWarnings("UseDataClass")
-internal class ApiClient
+import androidx.annotation.VisibleForTesting
+import retrofit2.Call
+import retrofit2.HttpException
+import retrofit2.Response
+import retrofit2.Retrofit
+
+internal class ApiClient @VisibleForTesting constructor(
+    retrofit: Retrofit,
+    private val hostAppVersion: String,
+    private val requestExecutor: RetrofitRequestExecutor = RetrofitRequestExecutor(retrofit),
+    private val listingApi: ListingApi = retrofit.create(ListingApi::class.java)
+) {
+
+    constructor(
+        baseUrl: String,
+        rasAppId: String,
+        subscriptionKey: String,
+        hostAppVersion: String
+    ) : this(
+        retrofit = createRetrofitClient(
+            baseUrl = baseUrl,
+            rasAppId = rasAppId,
+            subscriptionKey = subscriptionKey
+        ),
+        hostAppVersion = hostAppVersion
+    )
+
+    suspend fun list(): List<ListingEntity> {
+        val request = listingApi.list(hostAppVersion = hostAppVersion)
+        return requestExecutor.executeRequest(request)
+    }
+}
+
+internal class RetrofitRequestExecutor(
+    retrofit: Retrofit
+) {
+
+    private val errorConvertor = retrofit.responseBodyConverter<ErrorResponse>(
+        ErrorResponse::class.java,
+        arrayOfNulls<Annotation>(0)
+    )
+
+    suspend fun <T> executeRequest(call: Call<T>): T {
+        val response = call.execute()
+
+        if (response.isSuccessful) {
+            return response.body()!! // Body can't be null if request was successful
+        } else {
+            val error = response.errorBody()!! // Error body can't be null if request wasn't successful
+            throw MiniAppHttpException(
+                response = response,
+                errorMessage = errorConvertor.convert(error)?.message
+                    ?: "No error message provided by server."
+            )
+        }
+    }
+}
+
+internal data class ErrorResponse(
+    val code: Int,
+    val message: String
+)
+
+/**
+ * Exception thrown when the Mini App API returns an error response.
+ * @param response Response from the server
+ * @param errorMessage Error message returned by the server
+ */
+class MiniAppHttpException(
```

Comment:
Yep, HttpException can't parse the error message returned in the JSON body under the `error` key. This will most likely be a more specific error message than just the HTTP status message.